csrc/quantization/vectorization_utils.cuh

```diff
@@ -0,0 +1,51 @@
+// vector_utils.cuh
+#pragma once
+#include "vectorization.cuh"
+
+namespace vllm {
+
+template <typename InT, typename OutT, typename VecOp, typename ScaOp>
+__device__ inline void vectorize_with_alignment(
+    const InT* in, OutT* out, int len, int tid, int stride,
+    VecOp&& vec_op,    // vec_n_t<InT,16> -> vec_n_t<OutT,16>
+    ScaOp&& sca_op) {  // InT -> OutT
+  constexpr int VEC = 16;
+  constexpr int WIDTH = VEC * sizeof(InT);  // 64 B
+  uintptr_t addr = reinterpret_cast<uintptr_t>(in);
+
+  int misalignment_offset = addr & (WIDTH - 1);       // addr % 64
+  int alignment_bytes = WIDTH - misalignment_offset;  // 64 - (addr % 64)
+  int prefix_elems = alignment_bytes & (WIDTH - 1);   // handle 64
+  prefix_elems /= sizeof(InT);
+  prefix_elems = min(prefix_elems, len);  // 0 ≤ prefix < 16
+
+  // 1. prefill the when it is unsafe to vectorize
+  for (int i = tid; i < prefix_elems; i += stride) {
+    sca_op(out[i], in[i]);
+  }
+
+  in += prefix_elems;
+  out += prefix_elems;
+  len -= prefix_elems;
+
+  int num_vec = len / VEC;
+  using vin_t = vec_n_t<InT, VEC>;
+  using vout_t = vec_n_t<OutT, VEC>;
+  auto* v_in = reinterpret_cast<const vin_t*>(in);
+  auto* v_out = reinterpret_cast<vout_t*>(out);
+
+  // 2. vectorize the main part
+  for (int i = tid; i < num_vec; i += stride) {
+    vout_t tmp;
+    vec_op(tmp, v_in[i]);
```

Comment:
Could this just do the for loop here and call the `vec_op`? Or maybe `vec_op` can have a default parameter value `DefaultVec<VEC_SIZE>{sca_op}` that loops the `sca_op`?

Reply:
DefaultVec<VEC_SIZE> looks better for me, fixed